src/Http/Middleware/CsrfProtectionMiddleware.php

```diff
@@ -102,30 +95,40 @@ public function __construct(array $config = [])
      */
     public function process(ServerRequestInterface $request, RequestHandlerInterface $handler): ResponseInterface
     {
+        $method = $request->getMethod();
+        $hasData = in_array($method, ['PUT', 'POST', 'DELETE', 'PATCH'], true)
+            || $request->getParsedBody();
+
         if (
-            $this->whitelistCallback !== null
+            $hasData
+            && $this->whitelistCallback !== null
             && call_user_func($this->whitelistCallback, $request) === true
         ) {
+            $request = $this->_unsetTokenField($request);
+
             return $handler->handle($request);
         }
 
         $cookies = $request->getCookieParams();
         $cookieData = Hash::get($cookies, $this->_config['cookieName']);
 
         if ($cookieData !== null && strlen($cookieData) > 0) {
-            $request = $request->withAttribute('csrfToken', $cookieData);
+            $request = $request->withAttribute('_csrfToken', $cookieData);
         }
 
-        $method = $request->getMethod();
         if ($method === 'GET' && $cookieData === null) {
             $token = $this->_createToken();
-            $request = $request->withAttribute('csrfToken', $token);
+            $request = $request->withAttribute('_csrfToken', $token);
```

Comment:
Is this something we should note in the migration guide? I know some folks use csrftokens in javascript blocks so that xhr requests can have csrf tokens applied.

Instead of predicting this should we remove the prefix from this and _formToken?

Reply:
I am fine removing the prefix from `_formToken` instead. 

Though previously the token was set as a request param, so anyone using `$request->getParam('csrfToken')` will still have to update the call. But I guess if we keep the key same the update can be easily done with rector.

Reply:
I reverted prefixing the attribute name.

Reply:
Renaming methods based on the first parameter is something I've added to rector. I could add more patterns for csrfToken too.

Reply:
Okay, so this is good to merge?

Reply:
Sorry for bringing this up again but in 3.x the param name contain underscore prefix so I would prefer we keep it for attribute name too so `getParam('_csrfToken')` calls can be easily replaced with `getAttribute('_csrfToken')`.

I can add back prefix to `formToken` too in the form protector PR.